python@3.12: promote to default python3

branchvincent opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

---

Hopefully this is a smoother transition than last year since all formulae should now be using fully-qualified python binaries like `python3.11`

branchvincent commented:
So far without Linux
* bpytop: known https://github.com/Homebrew/homebrew-core/pull/121636#issuecomment-1407815158, will look into fixing this 
* condure: seen before, will look into fixing this
* pyqt@5: #150515
* trino: known, https://github.com/trinodb/trino/issues/18983983

Bo98 commented:
Might be worth revision bumping so the user doesn't end up in a scenario where they have Python 3.12 already installed but no `python3`.

Bo98 commented on Formula/p/python@3.12.rb:
```diff
@@ -466,23 +469,20 @@ def sitecustomize
         #{opt_libexec}/bin
 
       You can install Python packages with
-        pip#{version.major_minor} install <package>
+        pip3 install <package>
       They will install into the site-package directory
         #{HOMEBREW_PREFIX}/lib/python#{version.major_minor}/site-packages
```
Does this work anymore with the externally managed thing?

iMichka replied:
Doc update is being worked on here: https://github.com/Homebrew/brew/pull/16643

iMichka replied:
Doc update is done. I went for a caveats cleanup: if people use pip incorrectly there will be a message about virtualenvs and the externally managed process IMHO.

onlined commented:
What is blocking this PR right now? I can lend a hand.

nickmccurdy commented:
This appears to still be good.

nickmccurdy commented:
This appears to still be good. I'm still not sure why we wouldn't want to migrate to the newest version, especially considering the typing improvements.

SMillerDev reviewed (changes requested):
Awaiting https://github.com/Homebrew/homebrew-core/pull/150390#discussion_r1354994844

cho-m commented:
> Awaiting [#150390 (comment)](https://github.com/Homebrew/homebrew-core/pull/150390#discussion_r1354994844)

Which may also depend on https://github.com/Homebrew/homebrew-core/issues/157500 if we decide to drop majority of `python-*` formulae.

iMichka commented:
Re-running CI to get a clearer view of what is still broken

cho-m commented:
> Re-running CI to get a clearer view of what is still broken

On macOS, `gtk-doc` is failing:
```
  ==> Testing gtk-doc
  ==> /usr/local/Cellar/gtk-doc/1.33.2_1/bin/gtkdoc-scan --module=test
  Failed to execute: /usr/local/Cellar/gtk-doc/1.33.2_1/bin/gtkdoc-scan
```

Probably needs a shebang fix (preferably using a stable path across migrations):
```sh
#!@@HOMEBREW_PREFIX@@/opt/python@3.12/libexec/bin/python3
```

alex commented:
This was not actually fixed.

branchvincent commented:
Oops, didn't mean to close it

branchvincent commented:
Previous run: https://github.com/Homebrew/homebrew-core/actions/runs/7952028325?pr=150390

On linux:

* `gtk-doc`: Fixed via #163303
* `libcap-ng`: `No module named 'capng'` (unrelated; seen previously, thought I fixed with #162479 but seems python bindings aren't being installed in the bottle)
* `llvm@13`: `No module named 'clang'`

We still need `revision` bumps and to fix `llvm@13`, but we should be good to merge after one more run

p-linnane commented:
I've made sure we have an available long timeout label if you want to push this through now @branchvincent.

cho-m commented on Formula/p/python@3.12.rb:
```diff
@@ -460,23 +464,12 @@ def sitecustomize
-      They will install into the site-package directory
-        #{HOMEBREW_PREFIX}/lib/python#{version.major_minor}/site-packages
-
-      tkinter is no longer included with this formula, but it is available separately:
-        brew install python-tk@#{version.major_minor}
```
Did we want to remove this caveat? There is also the `python-gdbm` caveat in `python@3.11` that didn't carry over to `python@3.12`.

`python-tk@3.12` and `python-gdbm@3.12` will remain as they are part of base Python but we just put them into separate formulae (given that Homebrew doesn't support sub/split-packages).

Not important to handle here but may want to add caveats in follow up.

p-linnane replied:
I agree that it makes sense to handle this in a followup syntax PR.

iMichka replied:
I removed that line because it now felt useless. We had it for a few years now I think. I hope our users have adjusted their workflows, and my hope was to stop telling obvious things in the caveats.
But if you think we need to add it back: this can be done in a syntax-only follow-up pull request.

cho-m replied:
It doesn't matter to me whether we keep or remove it.

Homebrew's way of separate formulae is a bit odd since we don't support sub/split-packages. Most repositories would either install everything together (e.g. Arch Linux) or use subpackages (which colocates the related packages that use the same source distribution).

cho-m commented:
Failures that can be followed up on:
- `libcap-ng` is probably installing Python bindings into `$HOMEBREW_PREFIX/lib/...` rather than `#{lib}/...` since we don't have a sandbox on Linux. This means it passes when built from source but will fail later on when installed from bottle.
- `profanity` is building with `python3-embed` so Python feature is disabled on macOS and linked to `python@3.11` on Linux. May be worth adding a major-versioned symlink to `libexec` so we can use `PKG_CONFIG_PATH` rather than an `inreplace` workaround (or manually passing in corresponding `python_CFLAGS`/`python_LIBS`).

p-linnane commented:
Thanks for all the work here @branchvincent!

branchvincent commented:
And thanks @iMichka for all the doc updates!
